File: Exercise4/Threshold.py
# -*- coding: utf-8 -*-
"""
Created on Sun Nov  5 22:09:48 2017

@author: zhangzhexi
"""

import cv2

def main():
    image = cv2.imread("Test_images/Lenna.png",1)
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    cv2.imshow('Input Image',gray)
    cv2.imwrite('Input_image.jpg',gray)
    
    threshold_value = 128
    dst,thresh = cv2.threshold(gray, threshold_value, 255, cv2.THRESH_TRUNC)
    cv2.imshow("Threshold Image", thresh)
    cv2.imwrite("Threshold_image.jpg",thresh)

#    Binary Threshold
    dst,thresh_binary = cv2.threshold(gray, threshold_value, 255, cv2.THRESH_BINARY)
    cv2.imshow("Binary threshold", thresh_binary)
    cv2.imwrite("Threshold_binary.jpg", thresh_binary)

#    Band Thresholding
    threshold1 = 27
    threshold2 = 125
    dst,binary_image_1 = cv2.threshold(gray, threshold1, 255, cv2.THRESH_BINARY)
    dst,binary_image_2 = cv2.threshold(gray, threshold2, 255, cv2.THRESH_BINARY_INV)
    band_thresholded_image = cv2.bitwise_and(binary_image_1,binary_image_2)
    cv2.imshow("Band Thresholding", band_thresholded_image)
    cv2.imwrite("Band_Thresholding.jpg", band_thresholded_image)

#   Semi Thresholding
    current_threshold = 128
    max_threshold = 255;
    dst,semi_thresholded_image = cv2.threshold(gray,current_threshold,max_threshold,cv2.THRESH_BINARY_INV|cv2.THRESH_OTSU)
    semi_thresholded_image = cv2.bitwise_and(gray,semi_thresholded_image)
    cv2.imshow("Semi Thresholding",semi_thresholded_image)
    cv2.imwrite("Semi_Thresholding.jpg",semi_thresholded_image)

#   Adaptive Thresholding
    adaptive_thresh = cv2.adaptiveThreshold(gray,255,cv2.ADAPTIVE_THRESH_GAUSSIAN_C,cv2.THRESH_BINARY,101,10)
    cv2.imshow("Adaptive Thresholding", adaptive_thresh)
    cv2.imwrite("Adaptive_Thresholding.jpg", adaptive_thresh)
    
    cv2.waitKey(0)
    cv2.destroyAllWindows()

if __name__ == '__main__':
  main()

File: Exercise4/Threshold.cpp
#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/highgui/highgui.hpp"
#include <iostream>
 
 // Author Rishab Shah

using namespace cv;
 
int main( int argc, char** argv )
{
   Mat src, gray, dst;

  // Load an image
  src = imread( argv[1], 1 );
 
  // Convert the image to Gray
  cvtColor( src, gray, CV_RGB2GRAY );
 
  namedWindow("Input Image", CV_WINDOW_AUTOSIZE );
  imshow("Input Image", src);

  /* 0: Binary,      1: Binary Inverted,      2: Threshold Truncated,      3: Threshold to Zero,      4: Threshold to Zero Inverted    */
  int threshold_type = 2;    // slider 1 [0, 4]
  int threshold_value = 128;  // slider 2 [0 255]

    threshold(gray, dst, threshold_value, 255, threshold_type );
    imshow("Thresholded Image", dst );
    
         int current_threshold = 128;
        int max_threshold = 255;
	
	//Binary Threshold
	Mat thresholded;
	threshold(gray, thresholded, current_threshold, max_threshold, THRESH_BINARY);
	imshow("Binary threshold",thresholded);	

	// Band thresholding
	Mat binary_image1,binary_image2,band_thresholded_image;
	int threshold1 = 27, threshold2 = 125;
	threshold(gray, binary_image1, threshold1, max_threshold, THRESH_BINARY);
	threshold(gray, binary_image2, threshold2, max_threshold, THRESH_BINARY_INV);
	bitwise_and( binary_image1, binary_image2, band_thresholded_image );
        imshow("Band Thresholding", band_thresholded_image);

	// Semi thresholding
	Mat semi_thresholded_image;
	threshold(gray, semi_thresholded_image, current_threshold, max_threshold, THRESH_BINARY_INV | THRESH_OTSU);
	bitwise_and( gray, semi_thresholded_image, semi_thresholded_image );
        imshow("Semi Thresholding", semi_thresholded_image);

	// Adaptive thresholding
	Mat adaptive_thresh;
	adaptiveThreshold(gray, adaptive_thresh, 255.0, ADAPTIVE_THRESH_GAUSSIAN_C, THRESH_BINARY, 101, 10 );
	imshow("Adaptive Thresholding", adaptive_thresh);

  waitKey( 0 );
}
